driver-core/src/main/com/mongodb/client/model/WindowedComputations.java

```diff
@@ -0,0 +1,769 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.client.model;
+
+import com.mongodb.annotations.Beta;
+import com.mongodb.client.model.Windows.Bound;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonDocument;
+import org.bson.BsonDocumentWriter;
+import org.bson.BsonType;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.conversions.Bson;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.mongodb.assertions.Assertions.isTrueArgument;
+import static com.mongodb.client.model.MongoTimeUnit.MONTH;
+import static com.mongodb.client.model.MongoTimeUnit.YEAR;
+import static org.bson.assertions.Assertions.notNull;
+
+/**
+ * Builders for {@linkplain WindowedComputation windowed computations} used in the
+ * {@link Aggregates#setWindowFields(Object, Bson, List) $setWindowFields} pipeline stage
+ * of an aggregation pipeline. Each windowed computation is a triple:
+ * <ul>
+ *     <li>A window function. Some functions require documents in a window to be sorted
+ *     (see {@code sortBy} in {@link Aggregates#setWindowFields(Object, Bson, List)}).</li>
+ *     <li>An optional {@linkplain Window window}, a.k.a. frame.
+ *     Specifying {@code null} window is equivalent to specifying an unbounded window,
+ *     i.e., a window with both ends specified as {@link Bound#UNBOUNDED}.
+ *     Some window functions require to specify an explicit unbounded window instead of specifying {@code null}.</li>
+ *     <li>A path to an output field to be computed by the window function over the window.</li>
+ * </ul>
+ * A windowed computation is similar to an {@linkplain Accumulators accumulator} but does not result in folding documents constituting
+ * the window into a single document.
+ *
+ * @mongodb.driver.manual meta/aggregation-quick-reference/#field-paths Field paths
+ * @since 4.3
+ */
+@Beta
+public final class WindowedComputations {
```

Comment:
Did you have any other ideas for the name of this class?  It's a mouthful, and I don't love using the word computations because, in the end, isn't pretty much everything a computation?  Maybe there's something in the server scope or design doc that would suggest an alternate name?

Reply:
I have not seen any special name for this construct in the server documents, and in MQL each "windowed computation" is a field of a document representing the value of the `output` field, so one might say this is an "output field", which is even how it is called in https://docs-mongodbcom-staging.corp.mongodb.com/docs/docsworker-xlarge/window-functions/reference/operator/aggregation/setWindowFields/#syntax. However, this is a non-descriptive and misleading term, because the output field path is the smallest part of the construct. Besides, we cannot call it simply `OutputField` because such a name implies nothing about the relation with window functions, so the tern would need to be made more specific somehow.

I think that "windowed computation" is quite a descriptive term and not that long. I can also imagine it being called "window function call", because it represents both the function to be called and its arguments. However, it also includes the field path where to put the result, and this may not be a good fit for the term "function call". That said, I am open to consider other names and in no way imply that this name is the best there can be, but it is the best I was able to come up with.